src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/ValuesSources/WeaklyCachedRecoverableValueSource.cs

```diff
@@ -45,118 +66,118 @@ public WeaklyCachedRecoverableValueSource(ValueSource<T> initialValue)
+        /// </summary>
+        private void UpdateWeakReferenceAndEnqueueSaveTask_NoLock(T instance)
         {
-            saveTask.SafeContinueWith(t =>
-            {
-                using (Gate.DisposableWait(CancellationToken.None))
-                {
-                    // Only assume the instance is saved if the saveTask completed successfully. If the save did not
-                    // complete, we can still rely on a constant value source to provide the instance.
-                    _recoverySource = saveTask.Status == TaskStatus.RanToCompletion
-                        ? new AsyncLazy<T>(RecoverAsync, Recover, cacheResult: false)
-                        : new AsyncLazy<T>(instance);
-
-                    // Need to keep instance alive until recovery source is updated.
-                    GC.KeepAlive(instance);
-                }
-            }, TaskScheduler.Default);
-        }
+            Contract.ThrowIfTrue(Gate.CurrentCount != 0);
 
-        private Task? EnsureInstanceIsSavedAsync(T instance)
-        {
-            if (_weakReference == null)
-            {
-                _weakReference = new WeakReference<T>(instance);
-            }
-            else
-            {
-                _weakReference.SetTarget(instance);
-            }
+            _weakReference ??= new WeakReference<T>(instance);
+            _weakReference.SetTarget(instance);
 
+            // Ensure we only save once.
             if (!_saved)
             {
                 _saved = true;
                 using (s_taskGuard.DisposableWait())
                 {
-                    // force all save tasks to be in sequence so we don't hog all the threads
-                    s_latestTask = s_latestTask.SafeContinueWithFromAsync(t =>
-                         SaveAsync(instance, CancellationToken.None), CancellationToken.None, TaskScheduler.Default);
-                    return s_latestTask;
+                    // force all save tasks to be in sequence so we don't hog all the threads.
+                    s_latestTask = SaveAndResetInitialValue(s_latestTask);
                 }
             }
 
-#pragma warning disable VSTHRD114 // Avoid returning a null Task (False positive: https://github.com/microsoft/vs-threading/issues/637)
-            return null;
-#pragma warning restore VSTHRD114 // Avoid returning a null Task
+            return;
+
+            async Task SaveAndResetInitialValue(Task previousTask)
+            {
+                // First wait for the prior task in the chain to be done.  Ignore all errors from prior tasks.  They
+                // do not affect if we run or not.
+                await previousTask.NoThrowAwaitableInternal(captureContext: false);
+
+                // Now defer to our subclass to actually save the instance to secondary storage.
+                await SaveAsync(instance, CancellationToken.None).ConfigureAwait(false);
```

Comment:
This is now potentially running under that s_taskGuard if the previous await didn't yield -- are we OK with that? It seems like that means we're potentially blocking the thing asking for the value when the prior pattern of putting it into a separate queue really ensured we stayed reponsive?

Reply:
(note that gate is also going to be a synchronous block regardless which seems iffy...)

Reply:
Have updated to explicitly yield.  So, to me, this is a pref/style thing between ContinueWith vs async/await.  I *personally* struggle more with .ContinueWith as i find it harder to think about things like the exception semantics (e.g. when do or don't you want things to run afterwards).  I find the async/await approach nicer here where it's just mainline executino and things like NoThrowAwaitableInternal indicate we're swallowing, and normal await means exceptions prevent further statements from executing.

Reply:
I agree there's a stylistic/preference thing here absolutely; some really don't like ContinueWith. I think in this case though I really _don't_ like the change -- previously it was clear that there was a fire-into-a-queue and now this requires more lines of code to do, and we actually got it wrong the first time. More lines of code to write more bugs seems like this didn't move in the right direction IMHO. It's also sketchy looking since it looks like it's trying to lock around the entire operation when it's not -- that's clearer IMHO w/ ContinueWith too.

So I agree the code is correct functionally; I think the original was clearer though. I'll leave what to do to you.

Reply:
reasoning is fair.  i'll return it :)